File: fixtures/scale/dart/large/lib/ciphers/fnv1a/fnv1a_hash.dart
import '../../core/hash_function.dart';
import '../../util/bytes.dart';

/// FNV-1a 64-bit-ish: xor the byte, multiply by the prime, masked to 32 bits.
class Fnv1aHash implements HashFunction {
  @override
  String name() => 'fnv1a';

  @override
  int digest(String input) {
    var state = 0x1BF29CE4;
    for (final b in bytesOf(input)) {
      state = (state ^ b) & 0xFFFFFFFF;
      state = (state * 0x1000193) & 0xFFFFFFFF;
    }
    return state;
  }
}

File: fixtures/scale/kotlin/large/src/cryptobench/ciphers/fnv1a/Fnv1aHash.kt
package cryptobench.ciphers.fnv1a

import cryptobench.core.HashFunction
import cryptobench.util.Bytes

/** FNV-1a 64-bit: xor the byte, multiply by the prime. */
class Fnv1aHash : HashFunction {

    override fun name(): String = "fnv1a"

    override fun digest(input: String): Long {
        var state = -0x340d631b7bdddcdbL
        for (b in Bytes.of(input)) {
            state = state xor (b.toLong() and 0xFF)
            state *= 0x100000001B3L
        }
        return state
    }
}
